e3sm_diags/driver/utils/dataset_xr.py

```diff
@@ -720,26 +781,74 @@ def _get_dataset_with_derived_ts_var(self) -> xr.Dataset:
+        # Example: [xr.DataArray(name="PRECC",...), xr.DataArray(name="PRECL",...)]
+        # Unlike the climatology dataset, the source variables for time series
+        # data can be found in multiple datasets so a single xr.Dataset object
+        # is returned containing all of them.
+        src_var_keys = list(matching_target_var_map.keys())[0]
         ds = self._get_dataset_with_source_vars(src_var_keys)
 
-        # Get the source variable DataArrays.
-        # Example:
-        #   [xr.DataArray(name="PRECC",...), xr.DataArray(name="PRECL",...)]
-        src_vars = [ds[var] for var in src_var_keys]
+        # 3. Use the derivation function to derive the variable.
+        ds_final = self._get_dataset_with_derivation_func(
+            ds, derivation_func, src_var_keys, target_var
+        )
 
-        # Using the source variables, apply the matching derivation function.
-        derivation_func = list(matching_target_var_map.values())[0]
-        derived_var: xr.DataArray = derivation_func(*src_vars)
+        return ds_final
+
+    def _get_dataset_with_derivation_func(
+        self,
+        ds: xr.Dataset,
+        func: Callable,
+        src_var_keys: Tuple[str, ...],
+        target_var_key: str,
+    ) -> xr.Dataset:
+        """
+        Get the dataset with the target variable using the derivation function
+        and source variables.
 
-        # Add the derived variable to the final xr.Dataset object and return it.
-        ds[target_var] = derived_var
+        Parameters
+        ----------
+        ds : xr.Dataset
+            The dataset with source variables used for deriving the target
+            variable.
+        func : Callable
+            The derivation function that uses the source variables to derive
+            the target variables.
+        src_var_keys : Tuple[str, ...]
+            The source variable keys.
+        target_var_key : str
+            The target variable key.
 
-        return ds
+        Returns
+        -------
+        xr.Dataset
+            The dataset with the derived target variable.
+        """
+        func_args = [ds[var].copy() for var in src_var_keys]
+
+        # TODO: There is probably a cleaner way of determining which arguments
+        # to pass to which derivation functions. However, this requires
+        # extensive refactoring of the structure for derived variables (e.g.,
+        # the massive derived variables dictionary).
+        if func in FUNC_REQUIRES_DATASET_AND_TARGET_VAR:
+            func_args = [ds, target_var_key] + func_args  # type: ignore # pragma: nocover
+            ds_final = func(*func_args)  # pragma: nocover
+        elif func in FUNC_REQUIRES_TARGET_VAR:
+            func_args = [target_var_key] + func_args  # type: ignore # pragma: nocover
+            ds_final = func(*func_args)  # pragma: nocover
+        else:
+            derived_var = func(*func_args)
+            ds_final = ds.copy()
+            ds_final[target_var_key] = derived_var
```

Comment:
New block that passes arguments based on what derivation function.

The new `cosp_histogram_standardize()` and `cosp_bin_sum()` functions require `xr.Dataset` and/or `target_var_key` to add bounds and/or perform other operations.

Reply:
This block will be simplified a bit more since `cosp_histogram_standardize()` will no longer require the `xr.Dataset` object to add bounds to (this [comment](https://github.com/E3SM-Project/e3sm_diags/pull/748#discussion_r1471675469))

Simplified version:
```python
if func in FUNC_REQUIRES_TARGET_VAR:
    func_args = [target_var_key] + func_args  # type: ignore # pragma: nocover
    ds_final = func(*func_args)  # pragma: nocover
else:
    derived_var = func(*func_args)
    ds_final = ds.copy()
    ds_final[target_var_key] = derived_var
```